### PR TITLE
GH-37093: [Python] Add async Flight client with GetFlightInfo 

### LINKED ISSUE #37093: [FlightRPC][Python] Implement async GetFlightInfo
### Describe the enhancement requested

Add Python bindings to #36512.

### Component(s)

FlightRPC, Python

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

Demonstrate how to deal with async Flight from Python.

### What changes are included in this PR?

Add bindings for GetFlightInfo.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes, new APIs.

* Closes: #37093

### CODE REVIEW

**Comment by lidavidm** [comment]
CC @pitrou @jorisvandenbossche this is a skeleton of how to bind async C++ code to asyncio in Python, though I still need to get it working with both asyncio and Trio. (I'm looking at https://github.com/codypiersall/pynng/blob/master/pynng/_aio.py as a reference of how we might do that.)

File: python/pyarrow/_flight.pyx [hunk 1]
```diff
@@ -1219,6 +1224,80 @@ cdef class FlightMetadataWriter(_Weakrefable):
...
+    # headers/trailers - maybe a runtime option? since we'd have to copy all
+    # of them. or just always copy them? moot point for now since not yet
+    # exposed.
+    async def get_flight_info(
+        self,
+        descriptor: FlightDescriptor,
+        *,
+        options: FlightCallOptions = None,
+    ):
+        call = AsyncCall()
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:1268]: Could pass the asyncio loop to `AsyncCall` here.

File: python/pyarrow/_flight.pyx [hunk 2]
```diff
@@ -1219,6 +1224,80 @@ cdef class FlightMetadataWriter(_Weakrefable):
...
+        with nogil:
+            CAsyncGetFlightInfo(self.client.client.get(), deref(c_options), c_descriptor, call, callback)
+
+
+cdef void _client_async_get_flight_info(void* self, CFlightInfo* info, const CStatus& status):
+    cdef FlightInfo result = FlightInfo.__new__(FlightInfo)
+    call = <object> self
+    if not status.ok():
+        print("Failed")
+        call.exception = RuntimeError(status.ToString())
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:1290]: Should use our regular Status conversion routines.

File: python/pyarrow/_flight.pyx [hunk 3]
```diff
@@ -1320,6 +1399,11 @@ cdef class FlightClient(_Weakrefable):
             check_flight_status(CFlightClient.Connect(c_location, c_options
                                                       ).Value(&self.client))
 
+    def as_async(self) -> None:
+        if anyio is None:
+            raise RuntimeError("anyio is required for the async interface")
```
Comment (pitrou) [inline python/pyarrow/_flight.pyx:1404]: I'm not sure why we would want `anyio`. Just start with plain `asyncio`?

File: python/pyarrow/src/arrow/python/flight.cc [hunk 4]
```diff
@@ -383,6 +383,31 @@ Status CreateSchemaResult(const std::shared_ptr<arrow::Schema>& schema,
   return arrow::flight::SchemaResult::Make(*schema).Value(out);
 }
 
+void AsyncGetFlightInfo(
+    arrow::flight::FlightClient* client,
+    const arrow::flight::FlightCallOptions& options,
+    const arrow::flight::FlightDescriptor& descriptor,
+    PyObject* context, AsyncGetFlightInfoCallback callback) {
+  // TODO: OwnedRefNoGil the context?
```
Comment (pitrou) [inline python/pyarrow/src/arrow/python/flight.cc:391]: Should certainly do so indeed.

**Review by pitrou — COMMENTED** [review]
This looks like a good start to me.

File: python/pyarrow/_flight.pyx (hunk 3 above)
Comment (lidavidm) [reply python/pyarrow/_flight.pyx:1404]: The main thing is to support both asyncio and Trio users (though I think anyio is unnecessary and we can just directly support them both)

Comment (pitrou) [reply python/pyarrow/_flight.pyx:1404]: Why would we want that, though? asyncio is the de facto standard. And the anyio API for delegating stuff to external threads seems rather convoluted.

Comment (lidavidm) [reply python/pyarrow/_flight.pyx:1404]: I don't see a reason to just lock out a part of the ecosystem.

**Comment by lidavidm** [comment]
Updated, added basic tests. I'll deal with Trio in a later PR, as well as raising an RpcError-type exception instead of built in exceptions (again: RPC errors should be distinct, and should be handled uniformly), once I've reviewed the C++ side and decided I'm happy with TransportStatusDetail.

File: python/pyarrow/_flight.pyx [hunk 5]
```diff
@@ -1219,6 +1219,94 @@ cdef class FlightMetadataWriter(_Weakrefable):
...
+            raise self._exception
+        return self._result
+
+    def wakeup(self, *, exception=None, result=None) -> None:
+        """Finish the RPC call."""
+        if exception:
+            self._exception = exception
+        else:
+            self._result = result
+        # XXX: why does set() have to be done from within the loop?
```
Comment (lidavidm) [inline python/pyarrow/_flight.pyx:1251]: Ah, I need to dig into this still.

Comment (lidavidm) [reply python/pyarrow/_flight.pyx:1251]: Also, I wonder if we can just directly complete an asyncio.Future (though I am not sure if that generalizes to Trio, or to streaming RPCs)

File: python/pyarrow/_flight.pyx (hunk 3 above)
Comment (lidavidm) [reply python/pyarrow/_flight.pyx]: And FWIW, one of the very first issues opened against ADBC was request for async support via anyio specifically, to support both Trio and asyncio

**Comment by lidavidm** [comment]
Hmm, that's an interesting segfault in CI.

**Comment by lidavidm** [comment]
Alright:

- The weird `SystemError: <class 'SystemError'> returned a result with an error set` is inscrutable to me. It occurs along with a `TypeError: __traceback__ must be a traceback or None` message. `with_traceback` doesn't appear to be able to clear the problematic traceback and I can't get gdb to break on the source of the error properly.
- The segfault is also related to tracebacks, somehow; it reproduces with PYTHONDEVMODE and gdb shows it happens somewhere in `tb_traverse`, where `obj->ob_type` points to something that appears to have already been deallocated:

  ```
  #0  _PyObject_IS_GC (obj=0x7fff656ce7af) at /usr/local/src/conda/python-3.11.3/Include/object.h:772
  #1  visit_decref (op=0x7fff656ce7af, parent=parent@entry=0x7fff656ce710)
      at /usr/local/src/conda/python-3.11.3/Modules/gcmodule.c:452
  #2  0x000055555583df5b in tb_traverse (tb=0x7fff656ce710, visit=0x555555723a20 <visit_decref>, arg=0x7fff656ce710)
      at /usr/local/src/conda/python-3.11.3/Python/traceback.c:182
  ```

**Comment by lidavidm** [comment]
So, I don't understand why, but bouncing through a pure-Python thread fixes the segfault. It seems something about having the C-callback in the traceback at some point borks the interpreter. It does seem that the asyncio implementation tries to get the traceback unconditionally (in call_soon_threadsafe -> call_soon -> asyncio.events.Handle) so I suppose maybe Cython isn't giving a valid traceback object for `_client_async_get_flight_info`? (Should probably go confirm that.)

No clue about the first error...

**Comment by pitrou** [comment]
Can the segfault be reproduced locally?

**Comment by pitrou** [comment]
Ok, I'll take a look hopefully on Monday.

**Comment by lidavidm** [comment]
Looking at these:
- https://github.com/Delgan/loguru/issues/88
- https://github.com/cython/cython/issues/2735

It looks like Cython indeed doesn't generate stack frames for calls, and also modifying `wakeup` here to try to call `sys._getframe` like `asyncio.event.Handle` does crashes with `call stack is not deep enough`. So indeed, it seems we should avoid touching asyncio or generating exceptions from the native callback. Bouncing through a Python thread seems unavoidable from that perspective, unfortunately.

**Comment by lidavidm** [comment]
gRPC themselves do something quite different; I don't know if they ran into challenges like this, but they sidestep this neatly: https://ep2020.europython.eu/media/conference/slides/AfUnDYW-grpc-python-c-extensions-and-asyncio.pdf

> - gRPC Asyncio Python application would start a separated thread for polling gRPC events
> - This thread won’t use any Python object, during the polling
> - Events would be added into a C++ queue
> - Asyncio loop will be woken up by writing into a socket


(Note that grpcio in Python works quite differently and implements a lot of what gRPC-C++ provides itself, so that may also explain the architecture. In particular, they have to drive the gRPC event loop, while gRPC-C++ manages that for us.)

For now, though, I think the easy way would be to have a single-thread thread pool per client that's used to trigger these callbacks.

**Comment by lidavidm** [comment]
Our callback never raises an exception. However it may generate an exception that we raise from within the asyncio loop. (I guess asyncio must attach the 'bad' traceback to it somewhere, as context for debugging? Since if we don't generate an exception, things are fine.)

**Comment by pitrou** [comment]
> Our callback never raises an exception.

It does not do so deliberately, sure, but it would raise one unwillingly?

**Comment by lidavidm** [comment]
Hmm, maybe there is one and it's getting suppressed, or rather, Python trying to print the exception is what's crashing the interpreter?

<details>

```
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
PyFrame_GetCode (frame=frame@entry=0x7fff6a05257000) at /usr/local/src/conda/python-3.11.3/Objects/frameobject.c:1312
1312	/usr/local/src/conda/python-3.11.3/Objects/frameobject.c: No such file or directory.
(gdb) bt
#0  PyFrame_GetCode (frame=frame@entry=0x7fff6a05257000) at /usr/local/src/conda/python-3.11.3/Objects/frameobject.c:1312
#1  0x000055555569728e in tb_printinternal (margin=<optimized out>, indent=<optimized out>, limit=<optimized out>, 
    f=<optimized out>, tb=0x7ffff0ba8aef) at /usr/local/src/conda/python-3.11.3/Python/traceback.c:943
#2  _PyTraceBack_Print_Indented (v=v@entry=0x7fff6a3b0ff0, indent=0, margin=0x55555588988a "", 
    header_margin=<optimized out>, header=<optimized out>, f=0x7ffff705d8b0)
    at /usr/local/src/conda/python-3.11.3/Python/traceback.c:1018
#3  0x000055555568b9dc in print_exception_traceback (value=0x7fff684a9fd0, ctx=0x7fffffffd258)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:942
#4  print_exception (ctx=ctx@entry=0x7fffffffd258, value=value@entry=0x7fff684a9fd0)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1224
#5  0x000055555568b233 in print_exception_recursive (ctx=ctx@entry=0x7fffffffd258, value=value@entry=0x7fff684a9fd0)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1509
#6  0x000055555568b054 in _PyErr_Display (file=0x7ffff705d8b0, exception=<optimized out>, value=0x7fff684a9fd0, 
    tb=<optimized out>) at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1552
#7  0x000055555568acf1 in PyErr_Display (exception=0x555556230ba0, value=0x7fff684a9fd0, tb=0x7fff6a3b0ff0)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:1584
#8  0x000055555568abe7 in sys_excepthook_impl (traceback=<optimized out>, value=<optimized out>, 
    exctype=<optimized out>, module=<optimized out>) at /usr/local/src/conda/python-3.11.3/Python/sysmodule.c:745
#9  sys_excepthook (module=<optimized out>, args=0x7fffffffd380, nargs=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Python/clinic/sysmodule.c.h:73
#10 0x000055555575785b in cfunction_vectorcall_FASTCALL (func=func@entry=0x7ffff7250050, args=0x7fffffffd380, nargsf=3, 
    kwnames=<optimized out>) at /usr/local/src/conda/python-3.11.3/Objects/methodobject.c:427
#11 0x00005555557299b5 in _PyObject_VectorcallTstate (tstate=0x555555ad08b8 <_PyRuntime+166328>, 
    callable=0x7ffff7250050, args=<optimized out>, nargsf=<optimized out>, kwnames=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Include/internal/pycore_call.h:92
#12 0x0000555555828eae in _PyErr_PrintEx (tstate=0x555555ad08b8 <_PyRuntime+166328>, set_sys_last_vars=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:826
#13 0x000055555568a801 in PyErr_Print () at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:882
#14 _PyRun_SimpleFileObject (fp=<optimized out>, filename=<optimized out>, closeit=<optimized out>, 
    flags=<optimized out>) at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:446
#15 0x0000555555828a23 in _PyRun_AnyFileObject (fp=0x555555b225b0, filename=0x7ffff72672e0, closeit=1, 
    flags=0x7fffffffd498) at /usr/local/src/conda/python-3.11.3/Python/pythonrun.c:79
#16 0x0000555555822ab6 in pymain_run_file_obj (skip_source_first_line=0, filename=0x7ffff72672e0, 
    program_name=0x7ffff71e7040) at /usr/local/src/conda/python-3.11.3/Modules/main.c:360
#17 pymain_run_file (config=0x555555ab6900 <_PyRuntime+59904>) at /usr/local/src/conda/python-3.11.3/Modules/main.c:379
#18 pymain_run_python (exitcode=0x7fffffffd490) at /usr/local/src/conda/python-3.11.3/Modules/main.c:601
#19 Py_RunMain () at /usr/local/src/conda/python-3.11.3/Modules/main.c:680
#20 0x00005555557e5179 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Modules/main.c:734
#21 0x00007ffff7c58083 in __libc_start_main (main=0x5555557e50d0 <main>, argc=2, argv=0x7fffffffd6c8, 
    init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd6b8)
    at ../csu/libc-start.c:308
#22 0x00005555557e5015 in _start () at /usr/local/src/conda/python-3.11.3/Python/getplatform.c:9
```

</details>

**Comment by lidavidm** [comment]
Ah, _and_ the crash is now non-deterministic...

File: python/pyarrow/src/arrow/python/flight.cc [hunk 6]
```diff
@@ -383,6 +383,25 @@ Status CreateSchemaResult(const std::shared_ptr<arrow::Schema>& schema,
...
+  auto future = client->GetFlightInfoAsync(options, descriptor);
+  future.AddCallback([callback, py_context = std::move(py_context)](
+                         arrow::Result<arrow::flight::FlightInfo> result) {
+    std::ignore = SafeCallIntoPython([&] {
+      if (result.ok()) {
+        callback(py_context.obj(), &result.ValueOrDie(), result.status());
+      } else {
+        callback(py_context.obj(), nullptr, result.status());
+      }
+      return Status::OK();
```
Comment (pitrou) [inline python/pyarrow/src/arrow/python/flight.cc:400]: Perhaps this instead, to catch and bubble up the exception neatly?
```suggestion
      return CheckPyError();
```

Comment (lidavidm) [reply python/pyarrow/src/arrow/python/flight.cc:400]: I do have this locally.

Comment (lidavidm) [reply python/pyarrow/src/arrow/python/flight.cc]: I pushed that up

**Comment by lidavidm** [comment]
```
Fatal Python error: PyThreadState_Get: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
```

Hmm, am I doing something wrong on the C++ side?

**Comment by lidavidm** [comment]
Hmm, yeah.

```
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f25576b8859 in __GI_abort () at abort.c:79
#2  0x0000558134e9cf78 in fatal_error_exit (status=-1) at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2722
#3  fatal_error (fd=<optimized out>, header=header@entry=1, 
    prefix=prefix@entry=0x5581350bb890 <__func__.5.lto_priv.1> "_PyMem_DebugFree", 
    msg=msg@entry=0x55813504c2f0 "Python memory allocator called without holding the GIL", status=status@entry=-1)
    at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2833
#4  0x0000558134e9d334 in _Py_FatalErrorFunc (func=func@entry=0x5581350bb890 <__func__.5.lto_priv.1> "_PyMem_DebugFree", 
    msg=msg@entry=0x55813504c2f0 "Python memory allocator called without holding the GIL")
    at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2919
#5  0x0000558134e02b3c in _PyMem_DebugCheckGIL (func=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Objects/obmalloc.c:2683
#6  _PyMem_DebugFree (ctx=<optimized out>, ptr=<optimized out>)
    at /usr/local/src/conda/python-3.11.3/Objects/obmalloc.c:2707
#7  0x0000558134fe37f5 in _PyFaulthandler_Fini () at /usr/local/src/conda/python-3.11.3/Modules/faulthandler.c:1445
#8  0x0000558134e9d2ca in fatal_error (fd=2, header=header@entry=1, 
    prefix=prefix@entry=0x558135095510 <__func__.6.lto_priv.2> "PyThreadState_Get", 
    msg=msg@entry=0x55813504ddc0 "the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)", status=status@entry=-1) at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2891
#9  0x0000558134e9d334 in _Py_FatalErrorFunc (
    func=func@entry=0x558135095510 <__func__.6.lto_priv.2> "PyThreadState_Get", 
    msg=msg@entry=0x55813504ddc0 "the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)") at /usr/local/src/conda/python-3.11.3/Python/pylifecycle.c:2919
#10 0x0000558134e9d3e4 in _Py_FatalError_TstateNULL (
    func=func@entry=0x558135095510 <__func__.6.lto_priv.2> "PyThreadState_Get")
    at /usr/local/src/conda/python-3.11.3/Python/ceval.c:344
#11 0x0000558134e31174 in _Py_EnsureFuncTstateNotNULL (tstate=<optimized out>, func=<synthetic pointer>)
    at /usr/local/src/conda/python-3.11.3/Include/internal/pycore_pystate.h:94
#12 PyThreadState_Get () at /usr/local/src/conda/python-3.11.3/Python/pystate.c:1200
#13 subtype_dealloc (self=0x7f24c9ae4dd0) at /usr/local/src/conda/python-3.11.3/Objects/typeobject.c:1369
#14 0x00007f24c97aad52 in Py_DECREF (op=0x7f24c9ae4dd0)
    at /home/lidavidm/miniforge3/envs/dev/include/python3.11/object.h:538
#15 0x00007f24c97aad74 in Py_XDECREF (op=0x7f24c9ae4dd0)
    at /home/lidavidm/miniforge3/envs/dev/include/python3.11/object.h:602
#16 0x00007f24c97ab2e7 in arrow::py::OwnedRef::reset (this=0x55813774ba48, obj=0x0)
    at /home/lidavidm/Code/upstream/flight-async/python/pyarrow/src/arrow/python/common.h:150
#17 0x00007f24c97ab313 in arrow::py::OwnedRef::reset (this=0x55813774ba48)
    at /home/lidavidm/Code/upstream/flight-async/python/pyarrow/src/arrow/python/common.h:154
#18 0x00007f24c97a8e18 in operator() (__closure=0x55813774ba28, result=...)
    at /home/lidavidm/Code/upstream/flight-async/python/pyarrow/src/arrow/python/flight.cc:403
#19 0x00007f24c97aa7b4 in arrow::Future<arrow::flight::FlightInfo>::WrapResultyOnComplete::Callback<arrow::py::flight::AsyncGetFlightInfo(arrow::flight::FlightClient*, const arrow::flight::FlightCallOptions&, const arrow::flight::FlightDescriptor&, PyObject*, arrow::py::flight::AsyncGetFlightInfoCallback)::<lambda(arrow::Result<arrow::flight::FlightInfo>)> >::operator()(const arrow::FutureImpl &) (this=0x55813774ba28, impl=...)
    at /home/lidavidm/Code/upstream/flight-async/install/include/arrow/util/future.h:442
```

**Comment by pitrou** [comment]
You probably need to turn the `OwnedRef` into an `OwnedRefNoGIL`?


Le 10/08/2023 à 19:34, David Li a écrit :
> Hmm, yeah.

**Comment by lidavidm** [comment]
It *is* an OwnedRefNoGIL though (not sure why the backtrace is like that, presumably because release() is defined on the base class)

**Comment by lidavidm** [comment]
Ok, well setting that aside (just taking out the explicit release() seems to be fine), the issue is still that somehow, we're generating an exception with an invalid traceback, and when we either go to print that traceback out or garbage collect it, the interpreter dies. I wonder if it has something to do with us generating the exception from a cdef function + asyncio possibly doing something weird/unsafe with stackframes that doesn't mesh well with Cython cdef functions.

**Comment by lidavidm** [comment]
And then if I replace check_status with just a hardcoded Python exception, I still get a traceback with an invalid frame being garbage-collected somewhere.

**Comment by lidavidm** [comment]
Ok. I think I figured it out.

I forgot to Py_INCREF before using the OwnedRefNoGIL :facepalm:

File: python/pyarrow/_flight.pyx [hunk 7]
```diff
@@ -1219,6 +1220,93 @@ cdef class FlightMetadataWriter(_Weakrefable):
...
+cdef void _client_async_get_flight_info(void* self, CFlightInfo* info, const CStatus& status) except *:
+    """Bridge the C++ async call with the Python side."""
+    cdef:
+        FlightInfo result = FlightInfo.__new__(FlightInfo)
+    call: AsyncioCall = <object> self
+    if status.ok():
+        result.info.reset(new CFlightInfo(move(deref(info))))
+        call.wakeup(result=result)
+    else:
+        call.wakeup(exception=convert_status(status))
```
[the remark this answers is not in the document]
Comment (pitrou) [reply python/pyarrow/_flight.pyx:1307]: I've sketched something that seems to work here:
https://gist.github.com/pitrou/d4e5a386bc62d05fda6593f1b1e83a47

File: python/pyarrow/_flight.pyx [hunk 8]
```diff
@@ -1320,6 +1410,11 @@ cdef class FlightClient(_Weakrefable):
             check_flight_status(CFlightClient.Connect(c_location, c_options
                                                       ).Value(&self.client))
 
+    def as_async(self) -> None:
+        if self._pool is None:
+            self._pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+        return AsyncioFlightClient(self, self._pool)
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply python/pyarrow/_flight.pyx]: It's not, I forgot to remove this after trying something

File: python/pyarrow/_flight.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply python/pyarrow/_flight.pyx]: Thanks! Applied the patch + added an async.h as noted

**Comment by lidavidm** [comment]
Yes, this should be ready.

**Review by pitrou — APPROVED** [review]
Thanks a lot. This really opens interesting possibilities, I think.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by pitrou** [comment]
At least `test-debian-11-python-3` has failed with:
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=53319&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181&l=5448
```
usr/local/lib/python3.9/dist-packages/pyarrow/tests/test_flight_async.py:59: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/lib/python3.9/asyncio/runners.py:44: in run
    return loop.run_until_complete(main)
usr/lib/python3.9/asyncio/base_events.py:642: in run_until_complete
    return future.result()
usr/local/lib/python3.9/dist-packages/pyarrow/tests/test_flight_async.py:56: in _test
    info = await async_client.get_flight_info(descriptor)
pyarrow/_flight.pyx:1281: in get_flight_info
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   pyarrow.lib.ArrowNotImplementedError: gRPC 1.40 or newer is required to use async
```

We could perhaps expose `supports_async()` in Python.

**Comment by pitrou** [comment]
That said, it would perhaps be nicer to expose the same error message in Python as in C++...

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

File: cpp/src/arrow/flight/client.h [hunk 9]
```diff
@@ -396,6 +396,13 @@ class ARROW_FLIGHT_EXPORT FlightClient {
   /// \brief Whether this client supports asynchronous methods.
   bool supports_async() const;
 
+  /// \brief Check whether this client supports asynchronous methods.
+  ///
+  /// This is like supports_async(), except that a detailed error message
+  /// is returned if async support is not available.  If async support is
+  /// available, this function returns successfully.
+  Status CheckAsyncSupport() const;
```
Comment (pitrou) [inline cpp/src/arrow/flight/client.h:404]: @lidavidm Does this API look ok?

**Comment by pitrou** [comment]
CI failures are unrelated.

**Pull request merged**